### PR TITLE
test: add tests for rollbacks of database scripts

### PR DESCRIPTION
**Pull request by miparnisari**
This is to help detect problems like this: https://github.com/openfga/openfga/pull/691

Tested manually:

```shell
[17/04/23 7:48:07] ~/GitHub/openfga (test-migrations) $  make start-postgres 
docker run -d --name postgres -p 5432:5432 -e POSTGRES_USER=postgres -e POSTGRES_PASSWORD=password postgres:14
c6c088272b07293dbc12f4611dedbfbc8307267ebda45fc36a5e4059a05fa3d6
> use 'postgres://postgres:password@localhost:5432/postgres' to connect to postgres
[17/04/23 7:48:11] ~/GitHub/openfga (test-migrations) $ ./openfga migrate --datastore-engine postgres --datastore-uri 'postgres://postgres:password@localhost:5432/postgres' --version 4
2023/04/17 19:48:15 current version 0
2023/04/17 19:48:15 migrating to 4
2023/04/17 19:48:15 migration done
[17/04/23 7:48:15] ~/GitHub/openfga (test-migrations) $ ./openfga migrate --datastore-engine postgres --datastore-uri 'postgres://postgres:password@localhost:5432/postgres' --version 4
2023/04/17 19:48:21 current version 3
2023/04/17 19:48:21 migrating to 4
2023/04/17 19:48:21 migration done
[17/04/23 7:48:21] ~/GitHub/openfga (test-migrations) $ ./openfga migrate --datastore-engine postgres --datastore-uri 'postgres://postgres:password@localhost:5432/postgres' --version 3
2023/04/17 19:48:24 current version 3
2023/04/17 19:48:24 migrating to 3
2023/04/17 19:48:24 nothing to do
[17/04/23 7:48:24] ~/GitHub/openfga (test-migrations) $ ./openfga migrate --datastore-engine postgres --datastore-uri 'postgres://postgres:password@localhost:5432/postgres' --version 2
2023/04/17 19:48:26 current version 3
2023/04/17 19:48:26 migrating to 2
2023/04/17 19:48:26 migration done
[17/04/23 7:48:26] ~/GitHub/openfga (test-migrations) $ ./openfga migrate --datastore-engine postgres --datastore-uri 'postgres://postgres:password@localhost:5432/postgres' --version 1
2023/04/17 19:48:28 current version 2
2023/04/17 19:48:28 migrating to 1
2023/04/17 19:48:28 migration done
[17/04/23 7:48:28] ~/GitHub/openfga (test-migrations) $ ./openfga migrate --datastore-engine postgres --datastore-uri 'postgres://postgres:password@localhost:5432/postgres' --version 0
2023/04/17 19:48:29 current version 1
2023/04/17 19:48:29 running all migrations
2023/04/17 19:48:29 migration done
[17/04/23 7:48:29] ~/GitHub/openfga (test-migrations) $ 
```

### CODE REVIEW

File: cmd/migrate.go [hunk 1]
```diff
@@ -116,26 +116,33 @@ func runMigration(_ *cobra.Command, _ []string) error {
 
 	goose.SetBaseFS(assets.EmbedMigrations)
 
-	if version > 0 {
+	if targetVersion > 0 {
```
Comment (rhamzeh) [inline cmd/migrate.go:119]: Can we add the shortest case first? And make the logs more consistent

E.g. always print the current version and always print what migrations are about to be run and when they are done, e.g.

```go
currentVersion, err := goose.GetDBVersion(db)
if err != nil {
	log.Fatal(err)
}
if targetVersion == 0 {
  log.Println("current version %d, running all migrations")
...
} else {
  log.Printf("current version %d, migrating to %d", currentVersion, targetVersion)
  targetInt64Version := int64(targetVersion)
  if targetInt64Version == currentVersion {
    log.Println("nothing to do")
  } else ...
  ...
  ...
}
```
log.Println("migration done")

File: pkg/storage/mysql/mysql_test.go [hunk 2]
```diff
@@ -25,6 +27,30 @@ func TestMySQLDatastore(t *testing.T) {
...
+	// starts the container and runs migration up to the latest migration version available
+	testDatastore := storagefixtures.RunDatastoreTestContainer(t, engine)
+
+	uri := testDatastore.GetConnectionURI()
+	ds, err := New(uri, sqlcommon.NewConfig())
+	require.NoError(t, err)
+	defer ds.Close()
+
+	// going from version 3 to 4 when migration #4 doesn't exist is a no-op
+	version := testDatastore.GetDatabaseVersion() + 1
```
Comment (adriantam) [inline pkg/storage/mysql/mysql_test.go:41]: Do we have control on the database version in the test?  (Reason I am asking is that how do we know that the current version is 4 and maximum version is 4 - which will result in a no-op).

Should we run downgrade until 0 (which you already do) and then run upgrade to maximum to ensure both upgrade and downgrade are tested?

File: cmd/migrate.go [hunk 3]
```diff
@@ -116,28 +116,33 @@ func runMigration(_ *cobra.Command, _ []string) error {
...
 			log.Fatal(err)
 		}
-
-		int64Version := int64(version)
-		if int64Version < currentVersion {
-			if err := goose.DownTo(db, migrationsPath, int64Version); err != nil {
+	} else {
+		log.Printf("migrating to %d", targetVersion)
+		targetInt64Version := int64(targetVersion)
+		if targetInt64Version < currentVersion {
```
Comment (adriantam) [inline cmd/migrate.go:134]: Will the most likely case be 
i) no-op
ii) upgrade
iii) downgrade
?

In this case, does it make sense to order the if statement according to this order to optimize path?

File: pkg/testfixtures/storage/storage.go [hunk 4]
```diff
@@ -10,6 +10,9 @@ type DatastoreTestContainer interface {
 	// GetConnectionURI returns a connection string to the datastore instance running inside
 	// the container.
 	GetConnectionURI() string
+
+	// GetDatabaseVersion returns the last migration applied (e.g. 3) when the container was created
+	GetDatabaseVersion() int64
```
Comment (jon-whit) [inline pkg/testfixtures/storage/storage.go:15]:
```suggestion
	GetDatabaseSchemaVersion() int64
```

File: pkg/storage/postgres/postgres_test.go [hunk 5]
```diff
@@ -26,6 +28,30 @@ func TestPostgresDatastore(t *testing.T) {
...
+	// starts the container and runs migration up to the latest migration version available
+	testDatastore := storagefixtures.RunDatastoreTestContainer(t, engine)
+
+	uri := testDatastore.GetConnectionURI()
+	ds, err := New(uri, sqlcommon.NewConfig())
+	require.NoError(t, err)
+	defer ds.Close()
+
+	// going from version 3 to 4 when migration #4 doesn't exist is a no-op
+	version := testDatastore.GetDatabaseVersion() + 1
```
Comment (jon-whit) [inline pkg/storage/postgres/postgres_test.go:42]: This behavior and comment is a bit misleading. If you try to migrate to revision 4 and it is an undefined revision, then that should return an error. There should be two cases worth testing

1. Omitting the revision defaults to latest.

   `openfga migrate --datastore-engine=postgres --datastore-uri="..."` (defaults to latest)

2. Specifying the revision explicitly should require a valid revision number.

    `openfga migrate --datastore-engine=postgres --datastore-uri="..." 10` (error, revision 10 is undefined)

    `openfga migrate --datastore-engine=postgres --datastore-uri="..." 3` (success)

File: pkg/storage/mysql/mysql_test.go (hunk 2 above)
Comment (miparnisari) [reply pkg/storage/mysql/mysql_test.go:41]: By default, when you create a test container, it runs all `up` migrations up to the latest one available.

So in the test i'm doing:
1. try to update to one version more -> no-op
2. run `down` until version 0 is reached

File: pkg/storage/postgres/postgres_test.go (hunk 5 above)
Comment (miparnisari) [reply pkg/storage/postgres/postgres_test.go:42]: > If you try to migrate to revision 4 and it is an undefined revision, then that should return an error.

If you're on version 3 and try to migrate to version 4, Goose doesn't do anything. It feels weird to introduce an error when there is none, I see it as a deletion where it's already deleted, in REST that results in a no-op as well

File: cmd/migrate.go (hunk 3 above)
Comment (miparnisari) [reply cmd/migrate.go:134]: I don't understand, what would we be optimizing?

Comment (adriantam) [reply cmd/migrate.go:134]: What I meant is that should we do something like this?

```
   if targetInt64Version == currentVersion {
     // no-op
   } else if targetInt64Version < currentVersion {
     //upgrade
   } else {
     //downgrade
   }
```
as downgrade seldomly happen (save a comparison or two).

Comment (miparnisari) [reply cmd/migrate.go:134]: Migrations are run once in a blue moon.. i don't think that would classify as an optimization

**Review by adriantam — APPROVED** [review]

File: pkg/storage/postgres/postgres_test.go (hunk 5 above)
Comment (miparnisari) [reply pkg/storage/postgres/postgres_test.go]: @jon-whit i'm going to merge this as-is but if you think we should introduce an error when users try to migrate from version X to version X+1 when the highest one available is X, please open an issue. Thanks!

**Pull request merged**